Library/Core/UnoCore/Source/Uno/Threading/Thread.uno

```diff
@@ -0,0 +1,183 @@
+    {
+        public ThreadStateException(String message) : base(message)
+        {
+        }
+    }
+
+    [DotNetType("System.Threading.Thread")]
+    [extern(UNIX) Require("Source.Declaration", "static void* _ThreadFunc(void* arg) { @{ThreadMain(Thread):Call((@{Thread}) arg)}; return NULL; }")]
+    [extern(WIN32) Require("Source.Declaration", "static DWORD WINAPI _ThreadFunc(LPVOID lpParam) { @{ThreadMain(Thread):Call((@{Thread}) lpParam)}; return 0; }")]
+    public sealed class Thread
+    {
+        extern(CPLUSPLUS) static ThreadLocal _currentThread;
+
+        extern(CPLUSPLUS) static Thread()
+        {
+            _currentThread = extern<ThreadLocal> "uCreateThreadLocal(NULL)";
+        }
+
+        extern(CPLUSPLUS) static void ThreadMain(Thread thread)
+        {
+            extern "uAutoReleasePool pool";
+            extern "uSetThreadLocal(@{_currentThread}, $0)";
+
+            try
+            {
+                thread._threadStart();
+            }
+            catch (Exception e)
+            {
+                // TODO: Use some kind of exception callback..
+                debug_log "Unhandled exception in thread: " + e;
+            }
+
+            extern "uRelease($0)";
+        }
+
+        extern(CPLUSPLUS) ThreadHandle _threadHandle;
+
+        readonly ThreadStart _threadStart;
+        bool _started;
+
+        public Thread(ThreadStart start)
+        {
+            if (start == null)
+                throw new ArgumentNullException(nameof(start));
+
+            _threadStart = start;
+        }
+
+        private Thread(bool started)
+        {
+            _started = started;
+        }
+
+        public bool IsBackground
+        {
+            get { return false; }
+            set
+            {
+                if (value != false)
+                    throw new NotImplementedException();
+            }
+        }
+
+        public void Start()
+        {
+            if (_started)
+                throw new ThreadStateException("Thread is running or terminated; it cannot restart.");
+
+            _started = true;
+
+            // This is a bit tricky. We need to retain *before* the thread starts,
+            // otherwise the thread might be deleted before it has started.
+            // To avoid leaks, ThreadMain() will release the thread.
+            extern "uRetain($$)";
+
+            if defined(UNIX)
+            {
+                if (extern<int> "pthread_create(&@{$$._threadHandle}, NULL, _ThreadFunc, (void*)$$)" != 0)
+                {
+                    extern "uRelease($$)";
+                    throw new InvalidOperationException("pthread_create() failed!");
+                }
+            }
+            else if defined(WIN32)
+            {
+                extern "@{$$._threadHandle} = ::CreateThread(NULL, 0, _ThreadFunc, (LPVOID)$$, 0, NULL)";
+
+                if (extern<bool>(_threadHandle) "!$0" )
+                {
+                    extern "uRelease($$)";
```

Comment:
Previously this would leak the Thread-object, but now we've added code to release it.